ForumProject/startups/tests.py

```diff
@@ -1,3 +1,167 @@
+        '''
+        # Create a CustomUser for authentication
+        cls.user = CustomUser.objects.create_user(
+            email='dummy@user.com',
+            first_name='John',
+            last_name='Doe',
+            phone_number='+3801234567',
+            password='password',
+            is_active=True
+        )
+        
+        # Get JWT token for the user
+        refresh = RefreshToken.for_user(cls.user)
+        cls.token = str(refresh.access_token)
+        
+        # Sample data for testing
+        cls.startup_data = {
+            'startup_name': 'Django Dribblers',
+            'startup_industry': 'IT',
+            'startup_phone': '+3801234567',
+            'startup_country': 'UA',
+            'startup_city': 'Lviv',
+            'startup_address': 'I Franka 123'
+        }
+    
+    
+    def setUp(self):
+        '''
+        Setup that is executed before each test case.
+        '''
+        # Set up the test client and authenticate with the token
+        self.client = APIClient()
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+    
+    
+    def create_startup(self, data):
+        '''
+        Helper method to create a Startup with given data.
+        '''
+        return self.client.post(reverse('startups:startup-list'), data, format='json')
+    
+    
+    def assert_failure(self, data, expected_objects=0):
+        '''
+        Helper method to verify the Startup creation fails as expected
+        '''
+        response = self.create_startup(data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(Startup.objects.count(), expected_objects)
+
+    @transaction.atomic
+    def test_ok_create_startup(self):
+        '''
+        Tested creation of one Startup.
+        '''
+        response = self.create_startup(self.startup_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(Startup.objects.count(), 1)
+        self.assertEqual(Startup.objects.first().startup_name, 'Django Dribblers')
+
+    @transaction.atomic
+    def test_ok_create_two_startups(self):
+        '''
+        Tested creation of two Startups
+        '''
+        # Create first Startup
+        response = self.create_startup(self.startup_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(Startup.objects.count(), 1)
+        
+        # Create second Startup
+        self.startup_data['startup_name'] = 'Python catchers'
+        response = self.create_startup(self.startup_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(Startup.objects.count(), 2)
+
+        self.assertEqual(Startup.objects.filter(startup_name='Django Dribblers').exists(), True)
+        self.assertEqual(Startup.objects.filter(startup_name='Python catchers').exists(), True)
```

Comment:
The modification of `self.startup_data` directly affects subsequent tests since `setUpTestData` is only run once per class. It’s safer to copy`self.startup_data` to a local variable within each test method to avoid side effects. E.g.
```
def test_ok_create_two_startups(self):
    '''
    Tested creation of two Startups with unique names.
    '''
    # Create first Startup
    response = self.create_startup(self.startup_data.copy())
    self.assertEqual(response.status_code, status.HTTP_201_CREATED)
    self.assertEqual(Startup.objects.count(), 1)

    # Create second Startup with a different name
    modified_data = self.startup_data.copy()
    modified_data['startup_name'] = 'Python catchers'
    response = self.create_startup(modified_data)
    self.assertEqual(response.status_code, status.HTTP_201_CREATED)
    self.assertEqual(Startup.objects.count(), 2)
```

Reply:
I understand that even if the `self.startup_data` persists with modifications, each test works with an independent test database, and Django transactions roll back changes at the end of each test case. For this reason I see no need to use deep copies. If you insist that the issue of avoidance of side effects still needs to be attended to, I suggest then to move definition of startup_data dictionary from setUpTestData function to setUp (use of deepcopy would make the code less readable, in my view, and maybe more resourcefull). Please let me know.

Reply:
The change, wich was provided by me,  ensures that the original `startup_data` is not altered, preserving the integrity of other tests that may rely on the original data
https://docs.djangoproject.com/en/5.0/topics/testing/tools/
![image](https://github.com/Project-Stage-Academy/UA-1155-bravo/assets/39273210/9aa92fd1-688e-4d84-a2f5-6e94fbe6329e)
We're using PostgreSQL with SQL transactions full support

Reply:
I have moved the startup_data to SetUp - this way it gets regenerated at the start of each test, thereby ensuring any subsequent tests do not get affected. I think keeping startup_data in setUpTestData (purpose of which is to create the data that is supposed to remain unchanged across all the tests) only for the sake of making deep copy thereof in almost every test is less preferred route.